fedimint-api/src/db/mod.rs

```diff
@@ -169,10 +175,86 @@ impl Drop for CommitTracker {
     }
 }
 
+struct IsolatedDatabaseTransaction<'isolated, 'parent: 'isolated> {
+    tx: &'isolated mut DatabaseTransaction<'parent>,
+    prefix: Vec<u8>,
+}
+
+impl<'isolated, 'parent: 'isolated> IsolatedDatabaseTransaction<'isolated, 'parent> {
+    pub fn new(
+        dbtx: &'isolated mut DatabaseTransaction<'parent>,
+        module_instance_id: ModuleInstanceId,
+    ) -> IsolatedDatabaseTransaction<'isolated, 'parent> {
+        let mut prefix_bytes = vec![MODULE_GLOBAL_PREFIX];
+        module_instance_id
+            .consensus_encode(&mut prefix_bytes)
+            .expect("Error encoding module instance id as prefix");
+        IsolatedDatabaseTransaction {
+            tx: dbtx,
+            prefix: prefix_bytes,
+        }
+    }
+}
+
+#[async_trait]
+impl<'isolated, 'parent> IDatabaseTransaction<'isolated>
+    for IsolatedDatabaseTransaction<'isolated, 'parent>
+{
+    async fn raw_insert_bytes(&mut self, key: &[u8], value: Vec<u8>) -> Result<Option<Vec<u8>>> {
+        let mut key_with_prefix = self.prefix.clone();
+        key_with_prefix.append(&mut key.to_vec());
+        self.tx
+            .raw_insert_bytes(key_with_prefix.as_slice(), value)
+            .await
+    }
+
+    async fn raw_get_bytes(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        let mut key_with_prefix = self.prefix.clone();
+        key_with_prefix.append(&mut key.to_vec());
+        self.tx.raw_get_bytes(key_with_prefix.as_slice()).await
+    }
+
+    async fn raw_remove_entry(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        let mut key_with_prefix = self.prefix.clone();
+        key_with_prefix.append(&mut key.to_vec());
+        self.tx.raw_remove_entry(key_with_prefix.as_slice()).await
+    }
+
+    async fn raw_find_by_prefix(&mut self, key_prefix: &[u8]) -> PrefixIter<'_> {
+        let mut prefix_with_module = self.prefix.clone();
+        prefix_with_module.append(&mut key_prefix.to_vec());
+        let raw_prefix = self
+            .tx
+            .raw_find_by_prefix(prefix_with_module.as_slice())
+            .await;
+        Box::new(raw_prefix.map(|pair| match pair {
+            Ok(kv) => {
+                let key = kv.0;
+                let stripped_key = &key[(self.prefix.len())..];
```

Comment:
observation: Assumes prefix-freeness of `ModuleInstanceId` encodings.

Reply:
How come? The `self.prefix` is the only thing we appended, and only thing we should remove. If we nest isolated classess each one will take care of its part of the prefix.

Reply:
Had we change the defintion of `PrefixIter` to return `&[u8]` we might get rid of the `to_vec()` and since we are probably going to decode to the destination soon anyway, the higher level code would not care.

Reply:
Is the concern with to_vec just that it is an extra copy? I tried changing this to &[u8] but ran into some lifetime issues, since the returned key/value pairs are temporary values with a lifetime of the existing function. I think I would need to use a Box here if we wanted to return &[u8]?